akka-projection-testkit/src/main/scala/akka/projection/testkit/TestProjection.scala

```diff
@@ -225,19 +322,52 @@ class TestSourceProvider[Offset, Envelope] private (
       else sourceEvents.concat(Source.maybe)
     }
 
+  override def source(offset: Supplier[CompletionStage[Optional[Offset]]])
+      : CompletionStage[akka.stream.javadsl.Source[Envelope, NotUsed]] = {
+    implicit val ec = akka.dispatch.ExecutionContexts.parasitic
+    source(() => offset.get().toScala.map(_.asScala)).map(_.asJava).toJava
+  }
+
   override def extractOffset(envelope: Envelope): Offset = extractOffsetFn(envelope)
 
   override def extractCreationTime(envelope: Envelope): Long = extractCreationTimeFn(envelope)
 }
 
 @ApiMayChange
-class TestInMemoryOffsetStore[Offset](implicit val system: ActorSystem[_]) {
+object TestInMemoryOffsetStore {
+
+  /**
+   * An in-memory offset store that may be used with a [[TestProjection]].
+   *
+   * @param system - an [[akka.actor.typed.ActorSystem]] to provide an execution context to asynchronous operations.
+   */
+  def apply[Offset](system: ActorSystem[_]): TestInMemoryOffsetStore[Offset] =
+    new TestInMemoryOffsetStore[Offset](system)
+
+  /**
+   * An in-memory offset store that may be used with a [[TestProjection]].
+   *
+   * @param system - an [[akka.actor.typed.ActorSystem]] to provide an execution context to asynchronous operations.
+   */
+  def create[Offset](system: ActorSystem[_]): TestInMemoryOffsetStore[Offset] = apply(system)
+}
+
+@ApiMayChange
+class TestInMemoryOffsetStore[Offset] private (system: ActorSystem[_]) {
   private implicit val executionContext: ExecutionContext = system.executionContext
 
   private var savedOffsets = List[(ProjectionId, Offset)]()
```

Comment:
Scary to not have any concurrency constructs in this. Maybe a projection instance itself may have enough guards in place, but can probably be accessed concurrently by the test itself. For this I think it's easiest to make all methods `synchronized`.

Reply:
I'll use `synchronized`.  I didn't immediately think it was an issue because saved offsets were only ever mutated when `saveOffset` was called, which I think is only called by a single thread at a time (at least in our impls?)